Add support for manual token-based authorization

rottencandy opened the pull request:
## Fixes 
<!-- For e.g Fixes: https://issues.redhat.com/browse/HAC-XXX -->
https://issues.redhat.com/browse/HAC-940

## Description
- Add support for manual token upload using SPI api.
- Enable access check for quay.io urls.
<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->


## Type of change
<!-- Please delete options that are not relevant. -->

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 

https://user-images.githubusercontent.com/20013884/175023137-bb359854-34c7-4867-a654-984d8e8214ac.mp4

<!-- If change affects UI in any way, tag relevant UX people and add screenshots/gifs  -->


## How to test or reproduce?
- Enter a private git repo or container image
- Select "Use a token instead" option
<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

<!-- - [ ] Test A -->
<!-- - [ ] Test B -->

<!-- **Test Configuration(s)**: -->


## Browser conformance: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge


<!-- ## Checklist: -->

<!-- 
- [ ] Code follows the style guidelines
- [ ] Self-reviewed the code
- [ ] Added comments in hard-to-understand areas
- [ ] Made corresponding changes to the documentation
- [ ] Changes generate no new warnings
- [ ] Added tests that prove this fix is effective or that the feature works
- [ ] New and existing unit tests pass locally with new changes
- [ ] Any dependent changes have been merged and published in downstream modules 
-->

rottencandy commented:
/retest

rottencandy commented:
/retest

rottencandy commented:
/cc @sahil143 @rohitkrai03

sahil143 commented on src/components/AddComponent/AuthTokenModal.tsx:
```diff
@@ -0,0 +1,120 @@
+      try {
+        const token = await getToken();
+        await fetch(
+          `https://spi-oauth-route-spi-system.apps.appstudio-stage.x99m.p1.openshiftapps.com/token/${namespace}/${accessTokenName}`,
```
storing spi api in a const would be better

sahil143 commented on src/components/AddComponent/AuthTokenModal.tsx:
```diff
@@ -0,0 +1,120 @@
+
+export const AuthTokenModal: React.FC<AuthTokenModalProps> = ({ onClose, accessTokenName }) => {
+  const [error, setError] = React.useState();
+  const { namespace } = useNamespace();
+  const {
+    auth: { getToken },
+  } = useChrome();
+
+  const submitToken = React.useCallback(
+    async (values: FormikValues) => {
+      try {
+        const token = await getToken();
+        await fetch(
+          `https://spi-oauth-route-spi-system.apps.appstudio-stage.x99m.p1.openshiftapps.com/token/${namespace}/${accessTokenName}`,
+          {
+            method: 'POST',
+            credentials: 'include',
+            headers: {
+              Authorization: `Bearer ${token}`,
+            },
+            body: JSON.stringify({
+              username: values.username,
+              // eslint-disable-next-line camelcase
+              access_token: values.token,
+            }),
+          },
+        );
```
you can encapsulate this in a hook which return spiUtils
something like 
```
const {uploadToken} = useSpiAPI()
uploadToken(accessTokenName);
```

rottencandy replied:
Updated.

sahil143 commented on src/components/AddComponent/GitAuthorization.tsx:
```diff
@@ -19,8 +21,9 @@ export const GitAuthorization: React.FC = () => {
+  const showModal = useModalLauncher();
 
-  const [authUrl, loaded] = useAccessTokenBindingAuth(url);
+  const [authUrl, loaded, accessTokenName] = useAccessTokenBindingAuth(url);
```
```suggestion
  const [{authUrl, accessTokenName}, loaded] = useAccessTokenBindingAuth(url);
```

sahil143 commented on src/components/AddComponent/SourceSection.tsx:
```diff
@@ -65,29 +64,17 @@ export const SourceSection: React.FC<SourceSectionProps> = ({ onSamplesClick })
+    setValidated(ValidatedOptions.default);
+    setHelpText('Validating...');
+    setHelpTextInvalid('');
+    setShowAuthorization(false);
```
already setting setShowAuthorization before if condition
```suggestion
 
```

rottencandy commented:
/retest

rottencandy commented:
/retest

rottencandy commented:
> @rottencandy , once a repo is validated and then I change the URL. It continues to show `Authorized Access`.

Was happening because the fetched binding for previous URL was set after the handler for new URL is run, and it overwrote the `authSecret`.

Updated to only set secret if url hasn't changed after call. I've fixed repo validation behaviour in #130

rottencandy commented:
/retest

rottencandy commented:
/retest

rohitkrai03 commented:
/retest

Katka92 commented:
/retest

rohitkrai03 commented:
/retest

rottencandy commented:
/retest

rottencandy commented:
/retest

rottencandy commented:
/retest

rottencandy commented:
cc @openshift/team-hacdev-ux

sahil143 commented:
/lgtm
